pkg/api/components.go

```diff
@@ -0,0 +1,259 @@
+	}
+}
+
+func appendComponentIfNotPresent(components []KubernetesComponent, component KubernetesComponent) []KubernetesComponent {
+	if component.Name != "" {
+		i := getComponentsIndexByName(components, component.Name)
+		if i < 0 {
+			return append(components, component)
+		}
+	}
+	return components
+}
+
+func getComponentsIndexByName(components []KubernetesComponent, name string) int {
+	for i := range components {
+		if components[i].Name == name {
+			return i
+		}
+	}
+	return -1
+}
+
+// assignDefaultComponentVals will assign default values to component from defaults, for each property in component that has a zero value
+func assignDefaultComponentVals(component, defaultComponent KubernetesComponent, isUpgrade bool) KubernetesComponent {
+	if component.Name == "" {
+		component.Name = defaultComponent.Name
+	}
+	if component.Enabled == nil {
+		component.Enabled = defaultComponent.Enabled
+	}
+	if !to.Bool(component.Enabled) {
+		return KubernetesComponent{
+			Name:    component.Name,
+			Enabled: component.Enabled,
+		}
+	}
+	if component.Data != "" {
+		return KubernetesComponent{
+			Name:    component.Name,
+			Enabled: component.Enabled,
+			Data:    component.Data,
+		}
+	}
+	for i := range defaultComponent.Containers {
+		c := component.GetContainersIndexByName(defaultComponent.Containers[i].Name)
+		if c < 0 {
+			component.Containers = append(component.Containers, defaultComponent.Containers[i])
+		} else {
+			if component.Containers[c].Image == "" || isUpgrade {
+				component.Containers[c].Image = defaultComponent.Containers[i].Image
+			}
+			if component.Containers[c].CPURequests == "" {
+				component.Containers[c].CPURequests = defaultComponent.Containers[i].CPURequests
+			}
+			if component.Containers[c].MemoryRequests == "" {
+				component.Containers[c].MemoryRequests = defaultComponent.Containers[i].MemoryRequests
+			}
+			if component.Containers[c].CPULimits == "" {
+				component.Containers[c].CPULimits = defaultComponent.Containers[i].CPULimits
+			}
+			if component.Containers[c].MemoryLimits == "" {
+				component.Containers[c].MemoryLimits = defaultComponent.Containers[i].MemoryLimits
+			}
+		}
+	}
+	for key, val := range defaultComponent.Config {
+		if component.Config == nil {
+			component.Config = make(map[string]string)
+		}
+		if v, ok := component.Config[key]; !ok || v == "" {
+			component.Config[key] = val
+		}
+	}
+	return component
+}
+
+func synthesizeComponentsConfig(components []KubernetesComponent, defaultComponent KubernetesComponent, isUpgrade bool) {
+	i := getComponentsIndexByName(components, defaultComponent.Name)
+	if i >= 0 {
+		components[i] = assignDefaultComponentVals(components[i], defaultComponent, isUpgrade)
+	}
+}
+
+func getAPIServerDefaultCommandString(cs *ContainerService) string {
+	if common.IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.17.0") {
+		return fmt.Sprintf("\"kube-apiserver\"")
```

Comment:
Shouldn't we just return strings directly from these helper functions instead of using `fmt.Sprintf`? These aren't actually using any string formatting logic.